OpenConfig Platform Transceiver

nakano-omw opened the pull request:
# OpenConfig support for Platform Transceiver.

https://github.com/sonic-net/SONiC/pull/1858
The HLD has already been reviewed in the UMF WG.

Add support for following features:
- serial-no
- connector-type
- vendor
- vendor-part
- date-code
- supply-voltage
- laser-temperature
- output-power
- input-power
- laser-bias-current

nakano-omw committed:
add OC Platform Transceiver

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

mssonicbld commented:
/azp run

nakano-omw committed:
translib: gofmt pfm_app.go

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw commented:
@hdwhdw @anders-nexthop
This is the updated PR that we requested at the Extend SONiC Workshop. The HLD has already been reviewed in the UMF WG. I fixed the issues from the previous PR (https://github.com/sonic-net/sonic-mgmt-common/pull/160) and recreated this one so that the diff only includes the actual changes. Could you please review this version when you get a chance? Thanks!

hdwhdw commented on translib/pfm_app.go:
```diff
@@ -430,66 +534,1564 @@ func (app *PlatformApp) getSysEepromFromDb(eeprom *ocbinds.OpenconfigPlatform_Co
+	}
+
+	return nil
+}
```
I think this is valid.

nakano-omw commented:
@hdwhdw I'm currently reviewing the comments and working on the code fixes

hdwhdw commented:
@nakano-omw Thanks. I think there are unit test files https://github.com/sonic-net/sonic-mgmt-common/blob/master/translib/pfm_app_test.go so we can take advantage of that. Also feel free to break these long logic into separate files.

nakano-omw committed:
translib/pfm_app: rename Vendor_* struct fields to PascalCase

Address review comments on PR #201: the CompTransceiverStateDb struct
used snake_case fields (Vendor_Oui, Vendor_Rev, Vendor_Date) which
violate Go naming conventions. Rename to VendorOui, VendorRev,
VendorDate and update all references.

Also add an inline comment on the vendor-part code path documenting
that sourcing it from TRANSCEIVER_INFO.vendor_oui is intentional and
follows the HLD's DB mapping table (sonic-net/SONiC PR #1858,
Section 4 "Mapping between Openconfig YANG and Redis DB" Table 2),
to avoid future reviewers re-raising the question.

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
translib/pfm_app: collapse 8-lane channel structs into arrays

Address review comments on PR #201 (maintainer-endorsed "very valid"
on the output-power instance, plus systemic equivalents for
input-power and laser-bias-current):

  - CompTransceiverPhysicalChannelStateOutputPowerDb
  - CompTransceiverPhysicalChannelStateInputPowerDb
  - CompTransceiverPhysicalChannelStateLaserBiasCurrentDb

Each struct previously held eight identically-named float64 fields
(tx1power..tx8power, rx1power..rx8power, tx1bias..tx8bias). The
DbObj parser unrolled eight ParseFloat calls and the FromDb getter
unrolled an eight-case switch that differed only in the field name.

Replace each struct with a single fixed-size array of eight float64
values (TxPower / RxPower / TxBias). The parser becomes a for loop
that builds the field name with fmt.Sprintf, and the getter indexes
the array by laneIndex with a single bounds check. Net change is
-220 / +32 lines and removes the lowercase struct fields that also
violated Go naming conventions.

For LaserBiasCurrent, preserve the existing N/A -> NaN handling by
keeping the per-lane N/A check inside the loop, and switch from
strconv.ParseFloat("NaN", 64) to math.NaN() for clarity.

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
translib/pfm_app: rename threshold struct fields to PascalCase

Address review comment on PR #201: CompTransceiverThresholdStateDb
held 20 float64 fields with lowercase names (temphighalarm,
templowalarm, vcchighalarm, ..., txbiaslowwarning), which violate Go
naming conventions for struct fields. Rename them to PascalCase
(TempHighAlarm, TempLowAlarm, VccHighAlarm, ...) and update both the
parser assignments and the getter struct-field accesses.

The DB field name strings ("temphighalarm" etc.) are left unchanged
since they are the on-wire Redis keys defined by xcvrd, not Go
identifiers. Local variables inside the FromDb getter (used purely
as a stable address for a pointer take) also stay lowercase since
short-scoped locals do not need to follow exported-field
conventions.

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
translib/pfm_app_test: add transceiver unit-test scaffolding

Address review comment on PR #201 (maintainer-endorsed): the new
OpenConfig Platform Transceiver feature in translib/pfm_app.go had
no test coverage. Extend pfm_app_test.go with:

  - createTransceiverFactoryDb / clearTransceiverDataFromDb helpers
    that load and tear down STATE_DB fixtures for TRANSCEIVER_INFO,
    TRANSCEIVER_DOM_SENSOR and TRANSCEIVER_DOM_THRESHOLD using the
    same db.SetEntry pattern as the existing EEPROM_INFO fixture.

  - Test_PfmApp_TransceiverState which exercises the transceiver
    state subtree (serial-no, vendor, vendor-part, vendor-rev,
    date-code, connector-type, supply-voltage) and the CRITICAL /
    WARNING thresholds against a populated Ethernet0 transceiver
    fixture. Assertions are scoped to "GET returns without error",
    which is the regression guard for the handler reaching
    completion against valid input.

  - Test_PfmApp_TransceiverState_Missing which queries the same
    paths against a non-existent component (transceiver_EthernetMissing)
    and verifies the handler does not panic. This is a regression
    guard for the map-lookup nil-pointer concerns raised by
    Copilot's auto-review on PR #201 (the actual map-safety helpers
    will follow in a separate patch).

Fixtures are loaded inside each test and torn down via t.Cleanup so
the existing Test_PfmApp_TopLevelPath bulk EEPROM test is not
affected by run ordering. Strict JSON-equality assertions on the
transceiver subtree are intentionally deferred until the
encoding/precision of float and enum fields can be observed against
a real run.

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
translib/pfm_app: add map-lookup helpers and apply to parsers

Address review comments on PR #201 that flagged the raw
app.<X>Table[ifName].entry pattern at every DbObj parser entry
(serial-no, transceiver state, supply-voltage, laser-temperature,
output-power, input-power, laser-bias-current, threshold) and at
fourteen sites inside doGetPlatformInfo for applPortTable.

The cached maps store dbEntry values (not pointers), so a missing
key returns a zero-value dbEntry whose .entry.Field map is nil.
.Has and .Get on a nil Field map return false / "" rather than
panicking, so the existing code was not literally crashing -- but
silently treating "no entry" as "entry with every field absent"
made the missing-row case invisible at runtime and reads as
fragile to a reviewer.

Introduce four small helpers:

  getTransceiverInfoEntry(ifName)      -> (db.Value, bool)
  getTransceiverDomSensorEntry(ifName) -> (db.Value, bool)
  getTransceiverDomThresholdEntry(ifName) -> (db.Value, bool)
  getApplPortEntry(ifName)             -> (db.Value, bool)

Each wraps the underlying map indexing and reports whether the
interface was actually present. Update each get*DbObj parser to
call the helper, log a Warningf on missing entry, and short-circuit
with a sensible zero / NaN result. For applPortTable, all fourteen
doGetPlatformInfo call sites now go through getApplPortEntry; the
ok value is currently discarded because the existing follow-on
logic (.Get("lanes"), strings.Split, ...) is already missing-tolerant
and dedicated error handling for the empty/missing case is the
subject of a separate patch in this series.

The per-branch transceiver*Table lookups inside the get*FromDb
getters were not flagged by review and are left as-is in this patch;
they will go through the helper once the doGetPlatformInfo split
hoists the lookup into a single call per request.

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
translib: split transceiver getters into pfm_app_transceiver.go

Address review comment on PR #201: pfm_app.go had grown to roughly
2100 lines, dominated by the OpenConfig Platform Transceiver
getters that were appended for #201. Per the maintainer's note that
breaking the new logic into separate files is acceptable, move the
transceiver Db-struct types and getCompTransceiver* getters into a
new translib/pfm_app_transceiver.go in the same package.

What stays in pfm_app.go:
  - PlatformApp struct, init / translate* / process*
  - The four map-lookup helpers (getTransceiverInfoEntry,
    getTransceiverDomSensorEntry, getTransceiverDomThresholdEntry,
    getApplPortEntry) so callers across both files can share them
  - getCompStateDbObj / getCompStateFromDb (System Component state)
  - doGetPlatformInfo, which continues to be the single dispatch
    entry point and now calls into the transceiver getters
    located in the new file

What moves to pfm_app_transceiver.go:
  - CompTransceiverStateDb / CompTransceiverStateSupplyVoltageDb /
    CompTransceiverPhysicalChannelState{Laser{Temperature,
    BiasCurrent}, OutputPower, InputPower}Db /
    CompTransceiverThresholdStateDb
  - getCompTransceiverStateDbObj / getCompTransceiverStateFromDb
  - getCompTransceiverStateSupplyVoltage{DbObj, FromDb}
  - getCompTransceiverPhysicalChannelStateLaserTemperature{DbObj, FromDb}
  - getCompTransceiverPhysicalChannelStateOutputPower{DbObj, FromDb}
  - getCompTransceiverPhysicalChannelStateInputPower{DbObj, FromDb}
  - getCompTransceiverPhysicalChannelStateLaserBiasCurrent{DbObj, FromDb}
  - getCompTransceiverThresholdState{DbObj, FromDb}

This is purely a file-organisation change: no symbol is renamed,
no logic is altered. pfm_app.go drops from ~1990 to ~1440 lines,
and the new file is ~580 lines.

Drop the now-unused math, regexp and (already-unused) reflect-adjacent
imports from pfm_app.go's import block. pfm_app_transceiver.go has
its own minimal import block (fmt, math, regexp, strconv, strings,
ocbinds, glog).

doGetPlatformInfo itself remains long; further extraction by URI
pattern is out of scope for this patch and will be a follow-up.

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
translib/pfm_app_transceiver: log voltage/temperature parse errors

Address review comments on PR #201 (C07 line 689, C09 line 726):
SupplyVoltage and LaserTemperature DbObj parsers ignored the error
returned by strconv.ParseFloat, so a malformed STATE_DB value
(anything that is not "N/A" but also not a parseable float)
silently became the float64 zero value and was indistinguishable
from a real 0.0 V / 0.0 C reading.

Switch both parsers to the explicit form Copilot suggested: when
parsing fails, log a Warning with the offending value and store
math.NaN(). The "N/A" sentinel still maps directly to NaN as
before, matching the FromDb consumer's existing Has() guard.

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
translib/pfm_app: warn and skip on malformed lane id

Address review comments on PR #201 (C20 line 1326, C21 line 1327):
fourteen sites in doGetPlatformInfo iterate the comma-separated
APPL_DB PORT_TABLE.lanes value and call strconv.ParseUint(lane, 10, 16)
with the error discarded. On a malformed lane string the call
returns 0, which the existing code then maps to lane index 0 and
silently mis-attributes channel data to the wrong physical channel.
An empty PORT_TABLE.lanes value (strings.Split returns [""]) hits
the same path and ParseUint("") fails for the same reason.

Capture ParseUint's error into a fresh local parseErr (not the
function-level err, which the surrounding loop body uses to
propagate FromDb failures via break), log it with the offending
lane and ifName, and continue to the next entry. Empty-lanes is
implicitly covered by the same continue (ParseUint on "" fails).

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
translib/pfm_app: propagate LaserBiasCurrent FromDb error

Address review comment on PR #201 (C22): in doGetPlatformInfo the
ten call sites of getCompTransceiverPhysicalChannelStateLaserBiasCurrentFromDb
ignored the returned error, while the preceding calls to
*OutputPower* and *InputPower* checked the same return and broke
out of the enclosing loop. The asymmetry meant a bias-current
fetch failure would silently leave the per-lane oc_val partially
populated and let the outer for loop continue with stale or empty
data.

Mirror the existing pattern by checking err and breaking from the
enclosing for loop on failure, so all three physical-channel
getters now propagate errors consistently.

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
translib/pfm_app_transceiver: validate date-code via time.Parse

Address review comment on PR #201 (C05): the date-code extractor
applied a YYYY-MM-DD regex that constrained month to 01-12 and day
to 01-31 but did NOT validate the day against the actual length of
the month. Calendar impossibilities such as 2024-02-30 or 2024-04-31
passed validation and were emitted unchanged to oc_val.DateCode.

Replace the second regex-based validation with time.Parse on the
"2006-01-02" layout, which natively rejects month/day combinations
that do not exist. On success, reformat to the existing
RFC 3339-ish output (...T00:00:00.000Z); on failure log a warning
with the original string and skip the field, matching the prior
behavior of leaving DateCode unset for unparseable input.

The leading regex that extracts three digit groups from the raw
STATE_DB value is retained because xcvrd writes vendor_date with a
trailing time portion (e.g. "2024-01-01 00:00:00") and the
extraction is the simplest tolerant way to pick out the date prefix.

Signed-off-by: Kanji Nakano <kanji.nakano@ntt.com>

nakano-omw committed:
Merge branch 'master' into openconfig-tranceiver

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run
